### PR TITLE
[sw dv] Multi-site support for Verilator

### PR DESCRIPTION
**Pull request by martin-lueker**
- Support non-standard library & include paths via environment variables
  CFLAGS_FOR_BUILD and LDFLAGS_FOR_BUILD
- Add -D__STDC_CONSTANT_MACROS for versions of g++ which discard some
  C99 macros (such as UIINT64_C, which we use for some prince cipher code.)

Signed-off-by: Martin Lueker-Boden <martin.lueker-boden@wdc.com>

### CODE REVIEW

File: hw/top_earlgrey/dv/verilator/chip_sim.core
Comment (martin-lueker) [inline hw/top_earlgrey/dv/verilator/chip_sim.core:108]: Variables CFLAGS_FOR_BUILD and LDFLAGS_FOR_BUILD are chosen as this is what we use at WD for holding non-standard include paths for passing these with meson.  (In the case of meson/ninja builds we have to add custom paths and link-time parameters for using glibc-2.17 on RHEL6.)

Comment (martin-lueker) [reply hw/top_earlgrey/dv/verilator/chip_sim.core:108]: In this case we require this to point to `libelf` which for us is installed in a nonstandard location.
We could however add instead a call to pkg-config here.   Any preferences?

Comment (imphil) [reply hw/top_earlgrey/dv/verilator/chip_sim.core:108]: Instead of using `__STDC_CONSTANT_MACROS` it's probably better to remove UINT64_C in https://github.com/lowRISC/opentitan/blob/b9788b8ce1bf206f9d79c659b146af8cb63099f2/hw/ip/prim/dv/prim_prince/crypto_dpi_prince/prince_ref.h. Can you give that a try (or otherwise I'll have a look.)?

**Comment by imphil** [comment]
Thanks @martin-lueker. The solution you found is actually surprising and not really intended to work this way. (It's expanding the variables in the Makefile, which are more an implementation detail.) Can you add a comment e.g. "TODO: Variable expansion depends on edalize internals. Find better solution." to make us remember what's going on? I can't see a better solution at the moment without edalize/fusesoc changes, so let's not block on that.

File: hw/top_earlgrey/dv/verilator/chip_sim.core
Comment (martin-lueker) [reply hw/top_earlgrey/dv/verilator/chip_sim.core:108]: I can try this sure.  If this goes wrong would `ninja -C build-out test` find it?

Comment (imphil) [reply hw/top_earlgrey/dv/verilator/chip_sim.core:108]: No, the prim_prince DPI code is used by DV only. (private) CI should find it, or you can manually run a regression with dvsim. But UINT64_C is really rather useless in prince_ref.h, so you can simply remove it and everything should be OK.

**Comment by martin-lueker** [comment]
> TODO: Variable expansion depends on edalize internals. Find better solution.

Sure.  Will do.

File: hw/top_earlgrey/dv/verilator/chip_sim.core
Comment (martin-lueker) [reply hw/top_earlgrey/dv/verilator/chip_sim.core:108]: I'll just remove it then.  (FYI: @udinator, @rswarbrick, @msfschaffner as maintainers of this block)

**Review by imphil — APPROVED** [review]
Thanks!

File: hw/top_earlgrey/dv/verilator/chip_sim.core
Comment (imphil) [inline hw/top_earlgrey/dv/verilator/chip_sim.core:112]: Ah not quite. CI is failing with 

> verilator -f lowrisc_dv_chip_verilator_sim_0.1.vc --trace --trace-fst --trace-structs --trace-params --trace-max-array 1024 --unroll-count 512 -CFLAGS "-std=c++11 -Wall -DVM_TRACE_FMT_FST -DVL_USER_STOP -DTOPLEVEL_NAME=chip_sim_tb" -CFLAGS "" -LDFLAGS "-pthread -lutil -lelf" -LDFLAGS "" -Wall --threads 4 -Wno-fatal --threads 4
> %Error: Invalid option: -pthread -lutil -lelf

We probably need to pass CFLAGS and LDFLAGS only once, like this:

```suggestion
          # TODO: Variable expansion depends on edalize internals. Find better solution.
          #       (Applies to LDFLAGS expansion below as well)
          - '-CFLAGS "$(CFLAGS_FOR_BUILD) -std=c++11 -Wall -DVM_TRACE_FMT_FST -DVL_USER_STOP -DTOPLEVEL_NAME=chip_sim_tb"'
          - '-LDFLAGS "$(LDFLAGS_FOR_BUILD) -pthread -lutil -lelf"'
```

Comment (martin-lueker) [reply hw/top_earlgrey/dv/verilator/chip_sim.core]: Sounds good.  To confirm, that still works on our side also.

**Pull request merged**
